pkg/bootstrap/platform/discovery.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package platform
+
+import (
+	"sync"
+	"time"
+)
+
+const defaultTimeout = 5 * time.Second
+
+// Discover attempts to discover the host platform, defaulting to
+// `Unknown` if a platform cannot be discovered.
+func Discover() Environment {
+	return DiscoverWithTimeout(defaultTimeout)
+}
+
+// DiscoverWithTimeout attempts to discover the host platform, defaulting to
+// `Unknown` after the provided timeout.
+func DiscoverWithTimeout(timeout time.Duration) Environment {
+	plat := make(chan Environment, 2) // sized to match number of platform goroutines
+	done := make(chan bool)
+
+	var wg sync.WaitGroup
+	wg.Add(2) // check GCP and AWS
+
+	go func() {
+		if IsGCP() {
+			plat <- NewGCP()
+		}
+		wg.Done()
+	}()
+
+	go func() {
+		if IsAWS() {
+			plat <- NewAWS()
+		}
+		wg.Done()
+	}()
+
+	go func() {
+		wg.Wait()
+		done <- true
```

Comment:
Just sanity checking, is it possible for there to be a race when `done` is ahead of `plat`? Or is the write to channel+read in select synchronous so it will always be correct?

Reply:
I was worried about this race condition as well. I ran a ton of iterations of the same logic in a throwaway test, however, and could not trigger it. 

The contract for `select` is as:

> If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection. Otherwise, if there is a default case, that case is chosen. If there is no default case, the "select" statement blocks until at least one of the communications can proceed.

Looking at this again now, maybe the better thing to do is to surround the `Done()`s with `else` and then in a `defer` (or similar), just call `Done()` twice to ensure the goroutine is cleaned up.

Thoughts?

Reply:
Instead, added a nested select in `<- done` case that should quickly check the `plat` chan again to ensure no race issues are involved.